---
title: "Inden B cels | BCR Analysis | Gene Usage"
author: "Giorgio Gonnella"
date: "2024-10-14"
output:
  html_document:
    theme: sandstone
params:
  prjpath: "/srv/baia/prj/subclinical_dengue_scrnaseq/"
  libpath: "/home/giorgio/R-pkg-Seurat5/"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message=F, warning=F, echo=T)
.libPaths(c(params$libpath, .libPaths()))

library(Platypus)
library(ggplot2)

analysis_root <- paste0(params$prjpath, "analysis-1/")

input_step_dir <- paste0(analysis_root,
                         "steps/110.0.P.B.BCR_analysis.platypus/")
input_var_dir <- paste0(input_step_dir, "results/vars/")
if (!dir.exists(input_var_dir)) {
  stop("Input variable directory does not exist: ", input_var_dir)
}

this_step_dir <- paste0(analysis_root, "steps/110.0.V.B.BCR_results/")
if (!dir.exists(this_step_dir)) {
  stop("Current step directory does not exist: ", this_step_dir)
}
plots_dir <- paste0(this_step_dir, "results/plots/")
tables_dir <- paste0(this_step_dir, "results/tables/")
dir.create(plots_dir, showWarnings = F, recursive = T)
dir.create(tables_dir, showWarnings = F, recursive = T)

helpers_dir = paste0(params$prjpath, "helpers/")
source(paste0(helpers_dir, "ggplot2_helpers.R"))
source(paste0(helpers_dir, "platypus_helpers.R"))
```

# Load data

```{r load_vgm}
vgm <- readRDS(file = paste0(input_var_dir, "vgm.rds"))
vgm_HL <- readRDS(file = paste0(input_var_dir, "vgm_HL.rds"))
```

# Gene usage - Heavy chain

## C gene

### Isotype by sample

```{r}
data <- data.frame(
  condition = factor(vgm[["VDJ"]][["clinical_condition"]]),
  sample = factor(vgm[["VDJ"]][["sample"]]),
  gene = vgm[["GEX"]]$VDJ_cgene_grp
)
plot_data <- data %>%
  group_by(sample, condition, gene) %>%
  summarise(count = n())

# remove the NA values
plot_data <- plot_data %>% filter(!is.na(gene))

plot <- ggplot(plot_data, aes(x=gene, y=count, fill=condition)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("Isotype") +
  ylab("Count") +
  facet_wrap(~sample) +
  theme(plot.title.position = "plot", plot.title = element_text(hjust = 0.5)) +
  theme(legend.position="none") +
  scale_fill_manual(values=vgm[["GEX"]]@misc$cond_cols)
print(plot)
```

### Isotype by cell type

```{r}
data <- data.frame(
  condition = factor(vgm[["VDJ"]][["clinical_condition"]]),
  celltype = factor(vgm[["VDJ"]][["celltype"]]),
  gene = vgm[["GEX"]]$VDJ_cgene_grp
)
plot_data <- data %>%
  group_by(celltype, condition, gene) %>%
  summarise(count = n())

# remove the NA values
plot_data <- plot_data %>% filter(!is.na(gene))

plot <- ggplot(plot_data, aes(x=gene, y=count, fill=condition)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("Isotype") +
  ylab("Count") +
  facet_wrap(~celltype) +
  theme(plot.title.position = "plot", plot.title = element_text(hjust = 0.5)) +
  theme(legend.position="none") +
  scale_fill_manual(values=vgm[["GEX"]]@misc$cond_cols)
print(plot)
```

### Isotype on UMAP

```{r}
plot <- DimPlot(vgm[["GEX"]], reduction="umapharm", group.by="VDJ_cgene",
                split.by = 'clinical_condition',
                cols = vgm$GEX@misc$igh_gene_cols) +
                   ggtitle("UMAP by isotype")
split_print_legend(plot)
filename <- paste0(plots_dir, "isotype_umap.svg")
ggsave(filename, plot, device = "svg", width = 7, height = 5)
```

### Isotype distribution plot

```{r}
print(create_isotype_distribution_plot(vgm))
```

### Isotype distribution for single celltypes

```{r}
for (celltype in levels(vgm$VDJ$pruned.labels_singleR_fine)) {
  print(celltype)
  plot <- create_isotype_distribution_plot(vgm, sel_celltype = celltype,
                            celltype_col = "pruned.labels_singleR_fine")
  print(plot)
  celltype_sanitized <- gsub(" ", "_", celltype)
  filename <- paste0(plots_dir, "heavy_chain_Cgene_group_", celltype_sanitized,
                     ".svg")
  ggsave(filename, plot, device = "svg", width = 10, height = 10)
}
```

## V gene

### V gene group on UMAP

```{r}
plot <- DimPlot(vgm[["GEX"]], reduction="umapharm", group.by="VDJ_vgene_grp",
                split.by = 'clinical_condition') +
  ggtitle("Heavy chain V gene group")
split_print_legend(plot)
```

### V gene on UMAP

```{r}
plot <- DimPlot(vgm[["GEX"]], reduction="umapharm", group.by="VDJ_vgene1",
                split.by = 'clinical_condition') +
  ggtitle("Heavy chain V gene allele")
split_print_legend(plot)
```

## D gene

### D gene group on UMAP

```{r}
p<-DimPlot(vgm[["GEX"]], reduction="umapharm", group.by="VDJ_dgene_grp",
           split.by = 'clinical_condition') +
           ggtitle("Heavy chain D gene group")
split_print_legend(p)
```

### D gene on UMAP

```{r}
p<-DimPlot(vgm[["GEX"]], reduction="umapharm", group.by="VDJ_dgene1",
           split.by = 'clinical_condition') +
           ggtitle("Heavy chain D gene allele")
split_print_legend(p)
```

## J gene

### J gene on UMAP

```{r}
p<-DimPlot(vgm[["GEX"]], reduction="umapharm", group.by="VDJ_jgene1",
           split.by = 'clinical_condition') +
           ggtitle("Heavy chain J gene")
split_print_legend(p)
```

# Gene usage - Light chain

## C gene

### C gene locus on UMAP

```{r}
plot <- DimPlot(vgm_HL[["GEX"]], reduction="umapharm", group.by="VJ_cgene_grp",
                split.by = 'clinical_condition',
                na.value = "gray",
                cols =  c("Kappa" = "#008000", "Lambda" = "#ff8844")) +
  ggtitle("Light chain C gene group")
split_print_legend(plot)
```

### C gene locus by sample

```{r}
data <- data.frame(
  condition = factor(vgm_HL[["VDJ"]][["clinical_condition"]]),
  sample = factor(vgm_HL[["VDJ"]][["sample"]]),
  gene = vgm_HL[["GEX"]]$VJ_cgene_grp
)
plot_data <- data %>%
  group_by(sample, condition, gene) %>%
  summarise(count = n())

# remove the NA values
plot_data <- plot_data %>% filter(!is.na(gene))

plot <- ggplot(plot_data, aes(x=gene, y=count, fill=condition)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("Light chain C gene group") +
  ylab("Count") +
  facet_wrap(~sample) +
  theme(plot.title.position = "plot", plot.title = element_text(hjust = 0.5)) +
  theme(legend.position="none") +
  scale_fill_manual(values=vgm_HL[["GEX"]]@misc$cond_cols)
print(plot)
```

### C gene locus by cell type

```{r}
data <- data.frame(
  condition = factor(vgm_HL[["VDJ"]][["clinical_condition"]]),
  celltype = factor(vgm_HL[["VDJ"]][["celltype"]]),
  gene = vgm_HL[["GEX"]]$VJ_cgene_grp
)
plot_data <- data %>%
  group_by(celltype, condition, gene) %>%
  summarise(count = n())

# remove the NA values
plot_data <- plot_data %>% filter(!is.na(gene))

plot <- ggplot(plot_data, aes(x=gene, y=count, fill=condition)) +
  geom_bar(stat="identity", position="dodge") +
  xlab("Light chain C gene group") +
  ylab("Count") +
  facet_wrap(~celltype) +
  theme(plot.title.position = "plot", plot.title = element_text(hjust = 0.5)) +
  theme(legend.position="none") +
  scale_fill_manual(values=vgm_HL[["GEX"]]@misc$cond_cols)
print(plot)
```

### C gene allele on UMAP

```{r}
plot <- DimPlot(vgm_HL[["GEX"]], reduction="umapharm", group.by="VJ_cgene1",
                split.by = 'clinical_condition',
                na.value = "gray") +
  ggtitle("Light chain C gene allele")
split_print_legend(plot)
```

### C gene allele distribution plot

```{r}
print(light_chain_Cgene_distribution_plot(vgm_HL))
```

### C gene allele distribution for single celltypes

```{r}
for (celltype in levels(vgm_HL$VDJ$celltype)) {
  print(light_chain_Cgene_distribution_plot(vgm_HL, sel_celltype = celltype))
}
```

## V gene

### V gene group on UMAP

```{r}
plot <- DimPlot(vgm_HL[["GEX"]], reduction="umapharm", group.by="VJ_vgene_grp",
                split.by = 'clinical_condition',
                na.value = "gray") +
  ggtitle("Light chain V gene group")
split_print_legend(plot)
```

### V gene allele on UMAP

```{r}
plot <- DimPlot(vgm_HL[["GEX"]], reduction="umapharm", group.by="VJ_vgene1",
                split.by = 'clinical_condition',
                na.value = "gray") +
  ggtitle("Light chain V gene allele")
split_print_legend(plot)
```

## J gene

### J gene on UMAP

```{r}

plot <- DimPlot(vgm_HL[["GEX"]], reduction="umapharm", group.by="VJ_jgene1",
                split.by = 'clinical_condition',
                na.value = "gray") +
  ggtitle("Light chain J gene allele")
split_print_legend(plot)
```

## Differential gene pairings analysis

```{r}
output_file <- paste0(tables_dir, "BCR.genes_diff_usage.all.tsv")
write.table(vgm_HL$GEX@misc$gene_pairings$all, output_file, sep="\t",
            row.names = FALSE, quote = FALSE)

output_file <- paste0(tables_dir, "BCR.genes_diff_usage.signif.tsv")
write.table(vgm_HL$GEX@misc$gene_pairings$p_filt, output_file, sep="\t",
            row.names = FALSE, quote = FALSE)

output_file <- paste0(tables_dir, "BCR.genes_diff_usage.HPb.all.tsv")
write.table(vgm_HL$GEX@misc$gene_pairings$hpb_all, output_file, sep="\t",
            row.names = FALSE, quote = FALSE)

output_file <- paste0(tables_dir, "BCR.genes_diff_usage.HPb.signif.tsv")
write.table(vgm_HL$GEX@misc$gene_pairings$hpb_p_filt, output_file, sep="\t",
            row.names = FALSE, quote = FALSE)
```

### Gene usage pairings plots

```{r collect_signif_pairs_list}
signif_pairs = list()
results <- vgm_HL$GEX@misc$gene_pairings$p_filt
for (celltype in unique(results$celltype)) {
    signif_pairs[[celltype]] = results$genes[results$celltype == celltype]
}

signif_pairs_hpb = vgm_HL$GEX@misc$gene_pairings$hpb_p_filt$genes
```

```{r alluvial_plots_celltypes, fig.height=12, fig.width=8}
for (celltype in unique(vgm[["VDJ"]]$celltype)) {
  print(celltype)
  if (celltype %in% names(signif_pairs)) {
    plt <- gene_usage_alluvial(vgm_HL[["VDJ"]], "VDJ_vgene", "VJ_vgene",
                               left_lbl = "IGHV", right_lbl = "IGLV/IGKV",
                               sel_celltype = celltype,
                               celltype_col = "celltype",
                               grp1 = "Subclinical", grp2 = "Hospitalized",
                               min_clonotype_frequency=2,
                               left_lbl_pfx = "IGHV", right_lbl_pfx = "IG",
                               signif_pairs = signif_pairs[[celltype]],
                               stratum_label_size = 2.8,
                               stratum_label_threshold = 0.02,
                               stratum_label_size_prop_factor = 0)
    celltype_sanitized <- gsub("[^[:alnum:]]", "_", celltype)
    fname = paste0(plots_dir, "B.VH_VL_pairings.",
                   celltype_sanitized, ".svg")
    ggsave(fname, plt, device = "svg", width=6, height=9)
    print(plt)
  }
}
```

```{r alluvial_plot_unsplitted, fig.height=12, fig.width=8}
plt <- gene_usage_alluvial(vgm_HL[["VDJ"]], "VDJ_vgene", "VJ_vgene",
                           left_lbl = "IGHV", right_lbl = "IGLV/IGKV",
                           sel_celltype = "Plasmablasts",
                           celltype_col = "pruned.labels_singleR_fine",
                           grp1 = "Subclinical", grp2 = "Hospitalized",
                           min_clonotype_frequency = 2,
                           left_lbl_pfx = "IGHV", right_lbl_pfx = "IG",
                           signif_pairs = signif_pairs[["Plasmablasts"]],
                           stratum_label_size = 2.8,
                           stratum_label_threshold = 0.02,
                           stratum_label_size_prop_factor = 0)
fname = paste0(plots_dir, "B.VH_VL_pairings.Plasmablasts.svg")
ggsave(fname, plt, device = "svg", width=6, height=9)
print(plt)
```

```{r alluvial_plot_total, fig.height=12, fig.width=8}
plt <- gene_usage_alluvial(vgm_HL[["VDJ"]], "VDJ_vgene", "VJ_vgene",
                           left_lbl = "IGHV", right_lbl = "IGLV/IGKV",
                           grp1 = "Subclinical", grp2 = "Hospitalized",
                           left_lbl_pfx = "IGHV", right_lbl_pfx = "IG",
                           min_clonotype_frequency=2,
                           signif_pairs = signif_pairs$Total,
                           stratum_label_size=2.8,
                           stratum_label_threshold = 0.02,
                           stratum_label_size_prop_factor = 0)
fname = paste0(plots_dir, "B.VH_VL_pairings.Total.svg")
ggsave(fname, plt, device = "svg", width=6, height=9)
print(plt)
```

```{r, fig.height=18, fig.width=8}
plasmablasts <- subset(vgm_HL$VDJ,
                       celltype %in% c("Proliferating plasmablasts",
                                       "Non-proliferating plasmablasts"))
plt <- gene_usage_alluvial(plasmablasts, "VDJ_vgene", "VJ_vgene",
                           left_lbl = "IGHV", right_lbl = "IGLV/IGKV",
                           sel_celltype = "Hospitalized",
                           celltype_col = "clinical_condition",
                           grp_col = "pruned.labels_singleR_fine.mod",
                           grp1 = "Non-proliferating plasmablasts",
                           grp2 = "Proliferating plasmablasts",
                           min_clonotype_frequency=2,
                           left_lbl_pfx = "IGHV", right_lbl_pfx = "IG",
                           signif_pairs = signif_pairs_hpb,
                           stratum_label_size = 2.8,
                           stratum_label_threshold = 0.015,
                           stratum_label_size_prop_factor = 0)
fname = paste0(plots_dir, "B.VH_VL_pairings.HPb.svg")
ggsave(fname, plt, device = "svg", width=6, height=13.5)
print(plt)
```

# Session info

```{r}
sessionInfo()
```
